Undo environ modification in RequirementTracker cleanup

asottile opened the pull request.

cjerdonek commented:
Thanks for the report. This PR should have tests though (including e.g. to make sure it's not wiping out any previously set variable).

asottile commented:
@cjerdonek do you have a suggested file where to put tests for this in? I didn't see one for this file so I wasn't sure where to add them (it's currently untested 😬)

cjerdonek commented:
Try searching the code base for `RequirementTracker`, and you should find a test file called `test_req.py`. :)

asottile commented:
that file isn't really testing the `RequirementTracker` -- it is just using it.   Should I make a `test_req_tracker.py`?

cjerdonek commented:
> that file isn't really testing the RequirementTracker -- it is just using it.

That's fine. That's because there aren't any tests of it yet. That file seems to be testing a number of "req" related things if you look. For example, it includes `RequirementSet` tests even though there is a `req_set.py`.

asottile commented:
also fixed the context manager as well -- it was performing side-effects outside of  `__enter__`

cjerdonek commented on tests/unit/test_req.py:
```diff
@@ -636,3 +636,25 @@ def test_mismatched_versions(caplog, tmpdir):
+        with patch.dict(os.environ, clear=True):
+            yield
+
+    def test_no_side_effects_on_init(self):
```
I would name this something starting with `test_init` (to group tests by method being tested) and be more specific about what you're actually testing, e.g. `test_init__does_not_set_PIP_REQ_TRACKER`.

cjerdonek commented on tests/unit/test_req.py:
```diff
@@ -636,3 +636,25 @@ def test_mismatched_versions(caplog, tmpdir):
+        RequirementTracker()
+        assert 'PIP_REQ_TRACKER' not in os.environ
+
+    def test_keeps_original_environ(self, tmpdir):
```
How about `test_context_manager__with_PIP_REQ_TRACKER`?

cjerdonek commented on tests/unit/test_req.py:
```diff
@@ -636,3 +636,25 @@ def test_mismatched_versions(caplog, tmpdir):
+            assert os.environ['PIP_REQ_TRACKER'] == str(tmpdir)
+        assert os.environ['PIP_REQ_TRACKER'] == str(tmpdir)
+
+    def test_undoes_environ_set(self):
```
How about `test_context_manager__without_PIP_REQ_TRACKER`?

asottile replied:
then it's not clear what the test is doing, requiring a code comment -- I'd rather keep as much noise out of the test as possible -- your suggested name doesn't help indicate what the test is demonstrating

cjerdonek replied:
It tests entering and exiting, so I think the name should reflect that, with a focus on the `PIP_REQ_TRACKER`, and similarly for the other test.

cjerdonek commented on tests/unit/test_req.py:
```diff
@@ -636,3 +637,25 @@ def test_mismatched_versions(caplog, tmpdir):
+        assert os.environ['PIP_REQ_TRACKER'] == str(tmpdir)
+
+    def test_undoes_environ_set(self):
+        with RequirementTracker():
```
I would probably also `assert 'PIP_REQ_TRACKER' not in os.environ` before starting the manager just to be safe and so it's easier to see by looking at the test that it's working (even though not strictly necessary).

asottile replied:
there's an autouse fixture which clears the environ -- seems unnecessary to assert imo

cjerdonek commented on tests/unit/test_req.py:
```diff
@@ -636,3 +637,25 @@ def test_mismatched_versions(caplog, tmpdir):
+        os.environ['PIP_REQ_TRACKER'] = str(tmpdir)
+        with RequirementTracker():
+            assert os.environ['PIP_REQ_TRACKER'] == str(tmpdir)
+        assert os.environ['PIP_REQ_TRACKER'] == str(tmpdir)
```
Add a code comment noting that it's testing that the original value is preserved.

cjerdonek commented on tests/unit/test_req.py:
```diff
@@ -636,3 +637,25 @@ def test_mismatched_versions(caplog, tmpdir):
+    def test_undoes_environ_set(self):
+        with RequirementTracker():
+            assert 'PIP_REQ_TRACKER' in os.environ
+        assert 'PIP_REQ_TRACKER' not in os.environ
```
Add a code comment noting that the variable is cleaned up.

asottile replied:
the test name obviates the need for a comment here imo -- let me know if you feel strongly

cjerdonek replied:
I took the time to point you in the right direction for your PR, review your PR, and make the comments. I feel that all of the comments would improve the PR.

asottile replied:
I disagree that it's the right direction, cluttering a dead simple test with comments when the test name itself is clear seems worse.

I understand you took time to review the code and I thank you for that, but I really don't think the things you pointed out are worth another iteration.  Please push to my branch your suggestions if you feel strongly about them

cjerdonek commented on tests/unit/test_req.py:
```diff
@@ -636,3 +636,25 @@ def test_mismatched_versions(caplog, tmpdir):
     )
+
+
+class TestRequirementTracker(object):
```
That's not a reason to keep adding to the arbitrariness though. We can bring order as new things are added.

asottile commented:
@cjerdonek I'm sad that this took (essentially) a one line change and spiraled into a bikeshed naming fight -- can you please push your requested changes on top of my branch to move this forward?

benoit-pierre commented:
@asottile: you never mentioned why this is needed (although I agree it's cleaner): there's no use case where it actually matters in the pip code base.

benoit-pierre commented:
Well, using:
```python
from pip._internal import main
main(['install', 'six'])
main(['install', 'astpretty', '-v']))
```
is not supported (pip is a command line application, not a library).

I'd be interested to know which tests failed.

asottile commented:
> 
> 
> Well, using:
> 
> ```python
> from pip._internal import main
> main(['install', 'six'])
> main(['install', 'astpretty', '-v']))
> ```
> 
> is not supported (pip is a command line application, not a library).
> 
> I'd be interested to know which tests failed.

of course not -- but that's roughly what the failing tests did.  it's been weeks since I tracked it down and don't really have the effort to find them again so you'll have to trust me here ;)

cjerdonek commented:
@asottile I took time to review your changes, and instead of updating your patch, you argued against everything I suggested. In my experience, when I submit a PR and someone makes suggestions, I update my patch and things continue to proceed. People suggest nits all the time, and it's no problem. I just update my PR. Sometimes the suggestions are as trivial as removing a single line of whitespace or tweaking a word in a docstring. I've also never seen a submitter ask the reviewer to make the changes for them. Lastly, if there's a suggestion you don't 100% agree with, there's usually a way to update your patch to accommodate some of what they suggested. Most suggestions have a kernel of truth to them, and there can be more than one way to address their concern.

asottile commented:
@cjerdonek I felt that some of the suggestions were unfair, I'm being held to a much higher standard than the existing code.  I fixed two correctness issues, added tests for untested code and I was met with the following:

- introduce an arbitrary ordering to a test file which is unordered (I don't think this improves the patch whatsoever, I'd like to dismiss it)
- add `None` attributes to a context manager in `__init__` (I explained why I don't think this is the right thing to do -- it enables a certain class of programming errors)
- Rename the tests and add comments (the test names already describe the behaviour, I don't agree with intentionally making the test names worse and then requiring a comment so the intent is understood -- the well-named test functions already describe the desired behaviour).
  - First off, there is no suggestion in any contribution guidelines for the suggestion you're indicating
  - Second, if we look at the contents of the pip test suite, only 22 / 819 functions adhere to the style you are suggesting:
  ```
  $ git grep 'def test_' -- tests | wc -l
  819
  $ git grep 'def test_.*__' -- tests | wc -l
  22
  ```

Since I think all three of those actions would be either the wrong thing to do or unnecessary, I dismissed your review.  You've instead taken the stance that these are blocking despite my patch being made in good faith.  It's fairly common (and especially so with the [new github feature](https://help.github.com/articles/incorporating-feedback-in-your-pull-request/#applying-a-suggested-change)!) to push code to fix nitpicks.  I've even checked the "allow maintainers to push to my branch" option when creating the PR.  I've found it helps avoid discussions around nitpicks and allow things to move forward!  After all, I'm just trying to improve the software!

I hope you'll reconsider <3

cjerdonek commented:
With pip, I haven't noticed maintainers ever changing submitted PR's. But if you're telling me you're not willing to update your patch, then I can probably do so. It's easy enough.

asottile commented:
please do!

BrownTruck commented:
Hello!

I am an automated bot and I have noticed that this pull request is not currently able to be merged. If you are able to either merge the ``master`` branch into this pull request or rebase this pull request against ``master`` then it will be eligible for code review and hopefully merging!

BrownTruck commented:
Hello!

I am an automated bot and I have noticed that this pull request is not currently able to be merged. If you are able to either merge the ``master`` branch into this pull request or rebase this pull request against ``master`` then it will be eligible for code review and hopefully merging!

asottile commented:
good bot, I've resolved the conflict

BrownTruck commented:
Hello!

I am an automated bot and I have noticed that this pull request is not currently able to be merged. If you are able to either merge the ``master`` branch into this pull request or rebase this pull request against ``master`` then it will be eligible for code review and hopefully merging!

asottile commented:
thanks robot, I've resolved the conflict!

BrownTruck commented:
Hello!

I am an automated bot and I have noticed that this pull request is not currently able to be merged. If you are able to either merge the ``master`` branch into this pull request or rebase this pull request against ``master`` then it will be eligible for code review and hopefully merging!

asottile commented:
thanks robot, I've resolved the conflict!

BrownTruck commented:
Hello!

I am an automated bot and I have noticed that this pull request is not currently able to be merged. If you are able to either merge the ``master`` branch into this pull request or rebase this pull request against ``master`` then it will be eligible for code review and hopefully merging!

pradyunsg commented:
I'll pick this up from here. :)

pradyunsg commented:
Closing since I've picked this up in #7242, #7243 and potentially more follow ups.

pradyunsg commented:
Hurrah!

Thanks @asottile for finding this, filing this PR and for the discussions here. :)

Thanks @cjerdonek for the reviewing this PR. :)
